### PR TITLE
[HttpFoundation] Use convention to allow throwing from __toString()

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | - |
| License | MIT |
| Doc PR | - |

Corollary to #15076, works without it.

### CODE REVIEW

File: src/Symfony/Component/HttpFoundation/Request.php [hunk 1]
```diff
@@ -480,7 +480,7 @@ public function __toString()
         try {
             $content = $this->getContent();
         } catch (\LogicException $e) {
-            trigger_error($e->getMessage(), E_USER_ERROR);
+            return trigger_error($e, E_USER_ERROR);
```
Comment (Tobion) [inline src/Symfony/Component/HttpFoundation/Request.php]: trigger_error expects a string as first parameter and not an exception?! Also this changes the return value of the method. The method is supposed to return a string and not a boolean from trigger_error.

Comment (Tobion) [inline src/Symfony/Component/HttpFoundation/Request.php]: So this is a bc break.

**Comment by nicolas-grekas** [comment]
In any case, `trigger_error(..., E_USER_ERROR);` is the last line of the method, because E_USER_ERROR either kills the process or throws (which kills the process with a fatal error). But when an error handler like the one in #15076 is in place, it is possible to recover, and it's the only way to do that.

About the argument type, $e is automatically casted to a string (__toString() is in the Throwable interface) and this is the most accurate error message: the generated string contains the type of the exception, its message + stack trace excerpt. They are all required to debug correctly. $e->getMessage() is not enought.

File: src/Symfony/Component/HttpFoundation/Request.php (hunk 1 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/Request.php]: How that?

**Comment by Tobion** [comment]
It's not possible to gracefully handle an E_USER_ERROR and continue the programm?

**Comment by nicolas-grekas** [comment]
Technically it's possible with a custom error handler, but that would be wrong as it's not the semantic of U_USER_ERROR.

**Comment by Tobion** [comment]
The semantic of trigger_error is that it allows to continue the program flow of the method (which is the main difference to an exception). So `return trigger_error` breaks this contract and makes no sense to me.

File: src/Symfony/Component/HttpFoundation/Request.php (hunk 1 above)
Comment (Tobion) [inline src/Symfony/Component/HttpFoundation/Request.php]: It returns a completely different value obviously.

**Comment by nicolas-grekas** [comment]
This is not the semantic for E_USER_ERROR: http://3v4l.org/6rDU4

**Comment by nicolas-grekas** [comment]
Which is also why E_USER_ERROR (and E_RECOVERABLE ERROR) can't be made to continue, but are forced to throw:
https://github.com/symfony/symfony/blob/27e6f485c1fa0923ea8ff8be9bb64d9bce65c49e/src/Symfony/Component/Debug/ErrorHandler.php#L254

File: src/Symfony/Component/HttpFoundation/Request.php (hunk 1 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/Request.php]: This is always the last line (see https://github.com/symfony/symfony/pull/15077#issuecomment-114484146)

**Comment by Tobion** [comment]
> This is not the semantic for E_USER_ERROR: http://3v4l.org/6rDU4

But this is only the default error handler. A custom one can do things differently. And for those the behavior changes. This is probably an edge case, but for me this code stand-alone makes no sense.
I'm not against it, but it pretty much a hack which is bound to #15076. Without the debug error handler this change just creates an invalid toString() return value. That also means static code analysis will easily mark this as invalid.

So we can just as well provide a convenience method in the debug component for `return trigger_error` which is then used (which makes it a dependency).

File: src/Symfony/Component/HttpFoundation/Request.php (hunk 1 above)
Comment (Tobion) [inline src/Symfony/Component/HttpFoundation/Request.php]: it depends on which error handler is used.

**Comment by nicolas-grekas** [comment]
When looking for the perfect convention that would allow working around the restriction, I had in mind to not mandate a dependency on the debug component. I still thing this is very important.
The convention I propose in #15076, which is applied here, is to me the best. It allows for a enhanced experience when a non-compliant error handler is in use (native one or any other custom one), and opens for an even more enhanced experience when a compliant error handler is in use.

File: src/Symfony/Component/HttpFoundation/Request.php (hunk 1 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/Request.php]: You may be did not notice that the very line (the original `trigger_error($e->getMessage(), E_USER_ERROR);`) has been introduced a few days ago. It is not in any tagged version of Symfony. And the original intent and behavior (even before this very line was introduced itself) was really to stop the execution flow at this line.

**Comment by Tobion** [comment]
It's ugly but a sensible workaround. :+1:

**Comment by nicolas-grekas** [comment]
That's why I like this Debug component :)

**Pull request merged**